### ISSUE TITLE
Tracking Issue for allowing zero-sized memory accesses and offsets

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#117945]
This issue tracks implementing the t-opsem decision in https://github.com/rust-lang/unsafe-code-guidelines/issues/472. This will require adjustments in many places (codegen, Miri, library docs, reference, ...). The intention is to track here what needs to be done until the transition is complete.

- [x] update LLVM codegen
- [x] update cranelift codegen (not needed)
- [ ] update GCC codegen: deferred to https://github.com/rust-lang/rustc_codegen_gcc/issues/516
- [x] update Miri
- [x] update library docs
- [ ] update the reference: https://github.com/rust-lang/reference/pull/1541

### Implementation history

- https://github.com/rust-lang/rust/pull/117329
- https://github.com/rust-lang/rust/pull/124921

**Comment by bjorn3** [comment]
cg_clif accepts ZST memory accesses and pointer offsets already. Pointer offsets are implemented as integer addition which doesn't have UB and ZST memory accesses never get turned into loads and stores in cranelift ir as there is no instruction that does so.

**Comment by RalfJung** [comment]
> memory accesses never get turned into loads and stores in cranelift ir as there is no instruction that does so.

Besides direct accesses, the other concerns are the copy, write_bytes, compare_bytes intrinsics. Those must be implemented in a way that they are not UB when `elem_count*elem_size` is 0.

**Comment by bjorn3** [comment]
They are implemented by calling the respective libc functions which LLVM already expects to accept 0-sized accesses, right?

**Comment by RalfJung** [comment]
GCC codegen might also need updating, Cc @antoyo @GuillaumeGomez

**Comment by RalfJung** [comment]
> They are implemented by calling the respective libc functions which LLVM already expects to accept 0-sized accesses, right?

Well what LLVM assumes doesn't matter for the cranelift backend, does it? ;) But more importantly, Rust explicitly assumes this itself as [documented here](https://doc.rust-lang.org/nightly/core/index.html#how-to-use-the-core-library).

**Comment by GuillaumeGomez** [comment]
No problem. Please ping us when we need to update our part and thanks for the ping!

**Comment by RalfJung** [comment]
Well I'm asking you if you need to update anything. :) You need to make sure that the `Offset` MIR binop is compiled in a way that offset by 0 bytes is always Defined Behavior even if the pointer operand is null or dangling or out of bounds or whatever.

I think zero-sized memory accesses disappear in the SSA codegen infrastructure before your backend even sees them so they should be fine.

And finally the `copy`, `copy_nonoverlapping`, `write_bytes`, `compare_bytes` intrinsics need to be lowered in a way that they are Defined Behavior when the size is 0, even if the pointers are null or dangling or whatever.

**Comment by antoyo** [comment]
These intrinsics are implemented by calling the GCC builtin functions: memcmp, memset, memcpy, memmove.
I'll double-check, but it seems fine to have a count of zero, but not NULL pointers.

**Comment by RalfJung** [comment]
Okay, something needs to change then because we'll allow null pointers for the Rust intrinsics.

**Comment by RalfJung** [comment]
I updated https://github.com/rust-lang/rust/pull/117329 to make it ready to land.

@rust-lang/opsem to address [this concern](https://github.com/rust-lang/unsafe-code-guidelines/issues/472#issuecomment-1921686682), I made `offset_from` have library UB but not language UB on two pointers with the same address but provenance for different allocations. Please let me know what you think.

**Comment by RalfJung** [comment]
I created a reference PR at https://github.com/rust-lang/reference/pull/1541. That should complete the implementation of this feature. :)  (Aside from the GCC backend, which is tracked separately at https://github.com/rust-lang/rustc_codegen_gcc/issues/516.)
